radix-engine/src/track/db_key_mapper.rs

```diff
@@ -1,71 +1,147 @@
+///   makes this mapping effectively irreversible);
+/// - using a (shorter, but hard to crack) hash prefix for Substate key.
 pub struct SpreadPrefixKeyMapper;
 
 impl DatabaseKeyMapper for SpreadPrefixKeyMapper {
     fn to_db_partition_key(node_id: &NodeId, module_id: ModuleId) -> DbPartitionKey {
         let mut buffer = Vec::new();
         buffer.extend(node_id.as_ref());
         buffer.push(module_id.0);
-        let hash_bytes = hash(buffer).0[(Hash::LENGTH - 26)..Hash::LENGTH].to_vec(); // 26 bytes
+        let hash_bytes = hash(buffer).0[..Self::PARTITION_KEY_HASH_LENGTH].to_vec();
         DbPartitionKey(hash_bytes)
     }
 
     fn to_db_sort_key(key: &SubstateKey) -> DbSortKey {
-        let bytes = match key {
-            SubstateKey::Tuple(field) => {
-                vec![*field]
-            }
-            SubstateKey::Map(key) => {
-                hash(key).0[12..Hash::LENGTH].to_vec() // 20 bytes
-            }
-            SubstateKey::Sorted((bucket, key)) => {
-                let mut bytes = bucket.to_be_bytes().to_vec();
-                bytes.extend(hash(key).0[12..Hash::LENGTH].to_vec()); // 20 bytes
-                bytes
-            }
+        let (discriminator_byte, rest_bytes) = match key {
+            SubstateKey::Tuple(field) => (Self::TUPLE_DISCRIMINATOR, vec![*field]),
+            SubstateKey::Map(key) => (Self::MAP_DISCRIMINATOR, Self::to_hash_prefixed(key)),
+            SubstateKey::Sorted((bucket, key)) => (
+                Self::SORTED_DISCRIMINATOR,
+                [
+                    bucket.to_be_bytes().as_slice(),
+                    &Self::to_hash_prefixed(key),
+                ]
+                .concat(),
+            ),
         };
-        DbSortKey(bytes)
+        DbSortKey(combine(discriminator_byte, &rest_bytes))
+    }
+
+    fn from_db_sort_key(db_sort_key: &DbSortKey) -> SubstateKey {
+        let bytes = &db_sort_key.0;
+        let (discriminator_byte, rest_bytes) = (bytes[0], &bytes[1..]);
+        match discriminator_byte {
+            Self::TUPLE_DISCRIMINATOR => SubstateKey::Tuple(rest_bytes[0]),
+            Self::MAP_DISCRIMINATOR => {
+                SubstateKey::Map(Self::from_hash_prefixed(rest_bytes).to_vec())
+            }
+            Self::SORTED_DISCRIMINATOR => SubstateKey::Sorted((
+                u16::from_be_bytes(copy_u8_array(&rest_bytes[..2])),
+                Self::from_hash_prefixed(&rest_bytes[2..]).to_vec(),
+            )),
+            unexpected_byte => panic!("unexpected discriminator byte: {}", unexpected_byte),
+        }
+    }
+}
+
+impl SpreadPrefixKeyMapper {
+    /// Length of hashes that are used in the database *instead* of plain RE Node and Module IDs.
+    const PARTITION_KEY_HASH_LENGTH: usize = 26;
+
+    /// Discriminator bytes for telling apart different types of sort keys (needed for reversible
+    /// encoding).
+    const TUPLE_DISCRIMINATOR: u8 = 1;
```

Comment:
Slightly interesting choice to start discriminators from 0 but I don't mind it too much :D

Reply:
(coming from C/Java): it is a very good thing to see "illegal discriminator byte: 0" (instead of silent invalid result) when you try to parse a clear (zeroed) buffer :P